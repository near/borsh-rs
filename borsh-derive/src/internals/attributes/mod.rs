use syn::{Attribute, Path};

pub mod field;
pub mod item;
pub mod parsing;

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
pub struct Symbol {
    /// Name of the attribute
    pub name: &'static str,
    /// Expected value format representation for error printing
    pub expected: &'static str,
    support: AsyncSupport,
}

#[derive(Copy, Clone, PartialEq, PartialOrd, Eq, Ord)]
enum AsyncSupport {
    Sync,
    Async,
    Both,
}

impl Symbol {
    pub const fn new(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Both,
        }
    }

    pub const fn new_sync(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Sync,
        }
    }

    pub const fn new_async(name: &'static str, expected: &'static str) -> Self {
        Self {
            name,
            expected,
            support: AsyncSupport::Async,
        }
    }

    pub const fn test_support<const ASYNC_OR_SYNC_SWITCH: bool>(&self) -> bool {
        if ASYNC_OR_SYNC_SWITCH {
            matches!(self.support, AsyncSupport::Async | AsyncSupport::Both)
        } else {
            matches!(self.support, AsyncSupport::Sync | AsyncSupport::Both)
        }
    }
}

/// `borsh` - top level prefix in nested meta attribute
pub const BORSH: Symbol = Symbol::new("borsh", "borsh(...)");
/// `bound` - sub-borsh nested meta, field-level only; `BorshSerialize` and `BorshDeserialize` contexts
pub const BOUND: Symbol = Symbol::new_sync("bound", "bound(...)");
/// `async_bound` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
#[cfg(feature = "async")]
pub const ASYNC_BOUND: Symbol = Symbol::new_async("async_bound", "async_bound(...)");
///  `use_discriminant` - sub-borsh nested meta, item-level only, enums only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync` and `BorshDeserializeAsync` contexts
pub const USE_DISCRIMINANT: Symbol = Symbol::new("use_discriminant", "use_discriminant = ...");
/// `serialize` - sub-bound nested meta attribute
pub const SERIALIZE: Symbol = Symbol::new("serialize", "serialize = ...");
/// `deserialize` - sub-bound nested meta attribute
pub const DESERIALIZE: Symbol = Symbol::new("deserialize", "deserialize = ...");
/// `skip` - sub-borsh nested meta, field-level only attribute;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const SKIP: Symbol = Symbol::new("skip", "skip");
/// `init` - sub-borsh nested meta, item-level only attribute; `BorshDeserialize` and `BorshDeserializeAsync` contexts
pub const INIT: Symbol = Symbol::new("init", "init = ...");
/// `serialize_with` - sub-borsh nested meta, field-level only; `BorshSerialize` context
pub const SERIALIZE_WITH: Symbol = Symbol::new_sync("serialize_with", "serialize_with = ...");
/// `deserialize_with` - sub-borsh nested meta, field-level only; `BorshDeserialize` context
pub const DESERIALIZE_WITH: Symbol = Symbol::new_sync("deserialize_with", "deserialize_with = ...");
/// `serialize_with_async` - sub-borsh nested meta, field-level only; `BorshSerializeAsync` context
#[cfg(feature = "async")]
pub const SERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("serialize_with_async", "serialize_with_async = ...");
/// `deserialize_with_async` - sub-borsh nested meta, field-level only; `BorshDeserializeAsync` context
#[cfg(feature = "async")]
pub const DESERIALIZE_WITH_ASYNC: Symbol =
    Symbol::new_async("deserialize_with_async", "deserialize_with_async = ...");
/// `crate` - sub-borsh nested meta, item-level only;
/// `BorshSerialize`, `BorshDeserialize`, `BorshSerializeAsync`, `BorshDeserializeAsync` and `BorshSchema` contexts
pub const CRATE: Symbol = Symbol::new("crate", "crate = ...");

#[cfg(feature = "schema")]
pub mod schema_keys {
    use super::Symbol;

    /// `schema` - sub-borsh nested meta, `BorshSchema` context
    pub const SCHEMA: Symbol = Symbol::new("schema", "schema(...)");
    /// `params` - sub-schema nested meta, field-level only attribute
    pub const PARAMS: Symbol = Symbol::new("params", "params = ...");
    /// `serialize_with` - sub-borsh nested meta, field-level only, `BorshSerialize` context
    /// `with_funcs` - sub-schema nested meta, field-level only attribute
    pub const WITH_FUNCS: Symbol = Symbol::new("with_funcs", "with_funcs(...)");
    /// `declaration` - sub-with_funcs nested meta, field-level only attribute
    pub const DECLARATION: Symbol = Symbol::new("declaration", "declaration = ...");
    /// `definitions` - sub-with_funcs nested meta, field-level only attribute
    pub const DEFINITIONS: Symbol = Symbol::new("definitions", "definitions = ...");
}

#[derive(Clone, Copy)]
pub enum BoundType {
    Serialize,
    Deserialize,
}
impl PartialEq<Symbol> for Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

impl<'a> PartialEq<Symbol> for &'a Path {
    fn eq(&self, word: &Symbol) -> bool {
        self.is_ident(word.name)
    }
}

fn get_one_attribute(attrs: &[Attribute]) -> syn::Result<Option<&Attribute>> {
    let mut attrs = attrs.iter().filter(|attr| attr.path() == BORSH);
    let borsh = attrs.next();
    if let Some(other_borsh) = attrs.next() {
        return Err(syn::Error::new_spanned(
            other_borsh,
            format!("multiple `{}` attributes not allowed", BORSH.name),
        ));
    }
    Ok(borsh)
}
