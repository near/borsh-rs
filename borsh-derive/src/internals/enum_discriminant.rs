use std::{collections::HashMap, convert::TryFrom};

use proc_macro2::{Ident, TokenStream};
use quote::{quote, ToTokens};
use syn::{punctuated::Punctuated, token::Comma, Variant};

pub struct Discriminants(HashMap<Ident, TokenStream>);
impl Discriminants {
    /// Calculates the discriminant that will be assigned by the compiler.
    /// See: https://doc.rust-lang.org/reference/items/enumerations.html#assigning-discriminant-values
    pub fn new(variants: &Punctuated<Variant, Comma>) -> Self {
        let mut map = HashMap::new();
        let mut next_discriminant_if_not_specified = quote! {0};

        for variant in variants {
            let this_discriminant = variant.discriminant.clone().map_or_else(
                || next_discriminant_if_not_specified,
                |(_, e)| e.into_token_stream(),
            );

            next_discriminant_if_not_specified = quote! { #this_discriminant + 1 };
            map.insert(variant.ident.clone(), this_discriminant);
        }

        Self(map)
    }

    pub fn get(
        &self,
        variant_ident: &Ident,
        use_discriminant: bool,
        variant_idx: usize,
    ) -> syn::Result<TokenStream> {
        let variant_idx = u8::try_from(variant_idx).map_err(|err| {
            syn::Error::new(
                variant_ident.span(),
                format!("up to 256 enum variants are supported: {}", err),
            )
        })?;
        let result = if use_discriminant {
            self.0.get(variant_ident).unwrap().clone() // discriminant value
        } else {
            quote! { #variant_idx }
        };
        Ok(result)
    }
}
