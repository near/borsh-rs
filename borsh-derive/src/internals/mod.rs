pub mod attributes;
pub mod deserialize;
mod enum_discriminant;
mod generics;
#[cfg(feature = "schema")]
pub mod schema;
pub mod serialize;

pub mod cratename;

#[cfg(test)]
mod test_helpers;
