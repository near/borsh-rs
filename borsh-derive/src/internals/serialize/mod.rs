use std::convert::TryFrom;

use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{parse_quote, Expr, ExprPath, Generics, Ident, Index, Path, Type};

use super::generics;

pub mod enums;
pub mod structs;
pub mod unions;

struct GenericsOutput {
    overrides: Vec<syn::WherePredicate>,
    serialize_visitor: generics::FindTyParams,
}

impl GenericsOutput {
    fn new(generics: &Generics) -> Self {
        Self {
            overrides: vec![],
            serialize_visitor: generics::FindTyParams::new(generics),
        }
    }
    fn extend<const IS_ASYNC: bool>(self, where_clause: &mut syn::WhereClause, cratename: &Path) {
        let trait_path: Path = if IS_ASYNC {
            parse_quote! { #cratename::ser::BorshSerializeAsync }
        } else {
            parse_quote! { #cratename::ser::BorshSerialize }
        };
        let predicates =
            generics::compute_predicates(self.serialize_visitor.process_for_bounds(), &trait_path);
        where_clause.predicates.extend(predicates);
        where_clause.predicates.extend(self.overrides);
    }
}

pub enum FieldId {
    Struct(Ident),
    StructUnnamed(Index),
    Enum(Ident),
    EnumUnnamed(Index),
}

impl FieldId {
    fn index(field_idx: usize) -> syn::Result<Index> {
        let index = u32::try_from(field_idx).map_err(|err| {
            syn::Error::new(
                Span::call_site(),
                format!("up to 2^32 fields are supported {}", err),
            )
        })?;
        Ok(Index {
            index,
            span: Span::call_site(),
        })
    }
    pub fn new_struct_unnamed(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::StructUnnamed(index);
        Ok(result)
    }
    pub fn new_enum_unnamed(field_idx: usize) -> syn::Result<Self> {
        let index = Self::index(field_idx)?;
        let result = Self::EnumUnnamed(index);
        Ok(result)
    }
}

impl FieldId {
    fn serialize_arg(&self) -> Expr {
        match self {
            Self::Struct(name) => parse_quote! { &self.#name },
            Self::StructUnnamed(index) => parse_quote! { &self.#index },
            Self::Enum(name) => parse_quote! { #name },
            Self::EnumUnnamed(ind) => {
                let field = Ident::new(&format!("id{}", ind.index), Span::mixed_site());
                parse_quote! { #field }
            }
        }
    }

    /// function which computes derive output [proc_macro2::TokenStream]
    /// of code, which serializes single field
    pub fn serialize_output<const IS_ASYNC: bool>(
        &self,
        field_type: &Type,
        cratename: &Path,
        serialize_with: Option<ExprPath>,
    ) -> TokenStream2 {
        let arg: Expr = self.serialize_arg();
        let dot_await = IS_ASYNC.then(|| quote! { .await });
        if let Some(func) = serialize_with {
            quote! { #func(#arg, writer)#dot_await?; }
        } else {
            let serialize_trait = Ident::new(
                if IS_ASYNC {
                    "BorshSerializeAsync"
                } else {
                    "BorshSerialize"
                },
                Span::call_site(),
            );
            quote! { <#field_type as #cratename::#serialize_trait>::serialize(#arg, writer)#dot_await?; }
        }
    }

    pub fn enum_variant_header(&self, skipped: bool) -> Option<TokenStream2> {
        match self {
            Self::Struct(..) | Self::StructUnnamed(..) => unreachable!("no variant header"),
            Self::Enum(name) => (!skipped).then_some(quote! { #name, }),
            Self::EnumUnnamed(index) => {
                let field_ident = Ident::new(
                    &if skipped {
                        format!("_id{}", index.index)
                    } else {
                        format!("id{}", index.index)
                    },
                    Span::mixed_site(),
                );
                Some(quote! { #field_ident, })
            }
        }
    }
}
