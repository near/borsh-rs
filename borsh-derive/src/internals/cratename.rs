use proc_macro2::Span;
use proc_macro_crate::{crate_name, FoundCrate};
use syn::{Attribute, Error, Ident, Path};

use super::attributes::item;

pub(crate) const BORSH: &str = "borsh";

pub(crate) fn get(attrs: &[Attribute]) -> Result<Path, Error> {
    let path = item::get_crate(attrs)?;
    match path {
        Some(path) => Ok(path),
        None => {
            let ident = get_from_cargo();
            Ok(ident.into())
        }
    }
}

pub(crate) fn get_from_cargo() -> Ident {
    let name = &crate_name(BORSH).unwrap();
    let name = match name {
        FoundCrate::Itself => BORSH,
        FoundCrate::Name(name) => name.as_str(),
    };
    Ident::new(name, Span::call_site())
}
