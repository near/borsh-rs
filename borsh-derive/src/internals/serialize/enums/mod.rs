use cfg_if::cfg_if;
use proc_macro2::{Span, TokenStream as TokenStream2};
use quote::quote;
use syn::{Fields, Ident, ItemEnum, Lifetime, Path, Token, Variant};

use crate::internals::{
    attributes::{field, item, BoundType},
    enum_discriminant::Discriminants,
    generics, serialize,
};

pub fn process<const IS_ASYNC: bool>(
    input: ItemEnum,
    cratename: Path,
) -> syn::Result<TokenStream2> {
    let enum_ident = &input.ident;
    let use_discriminant = item::contains_use_discriminant(&input)?;
    let generics = generics::without_defaults(input.generics);
    let (impl_generics, ty_generics, where_clause) = generics.split_for_impl();
    let mut where_clause = generics::default_where(where_clause);
    let mut generics_output = serialize::GenericsOutput::new(&generics);
    let mut all_variants_idx_body = TokenStream2::new();
    let mut fields_body = TokenStream2::new();
    let discriminants = Discriminants::new(&input.variants);
    let mut has_unit_variant = false;

    for (variant_idx, variant) in input.variants.iter().enumerate() {
        let variant_ident = &variant.ident;
        let discriminant_value = discriminants.get(variant_ident, use_discriminant, variant_idx)?;
        let variant_output = process_variant::<IS_ASYNC>(
            variant,
            enum_ident,
            &discriminant_value,
            &cratename,
            &mut generics_output,
        )?;
        all_variants_idx_body.extend(variant_output.variant_idx_body);
        match variant_output.body {
            VariantBody::Unit => has_unit_variant = true,
            VariantBody::Fields(VariantFields { header, body }) => fields_body.extend(quote!(
                #enum_ident::#variant_ident #header => {
                    #body
                }
            )),
        }
    }
    let fields_body = optimize_fields_body(fields_body, has_unit_variant);
    generics_output.extend::<IS_ASYNC>(&mut where_clause, &cratename);

    let serialize_trait = Ident::new(
        if IS_ASYNC {
            "BorshSerializeAsync"
        } else {
            "BorshSerialize"
        },
        Span::call_site(),
    );
    let writer_trait = if IS_ASYNC {
        quote! { async_io::AsyncWrite }
    } else {
        quote! { io::Write }
    };
    let r#async = IS_ASYNC.then(|| Token![async](Span::call_site()));
    let lifetime = IS_ASYNC.then(|| Lifetime::new("'async_variant", Span::call_site()));
    let lt_comma = IS_ASYNC.then(|| Token![,](Span::call_site()));

    let write_variant_idx = if IS_ASYNC {
        quote! { writer.write_u8(variant_idx).await }
    } else {
        quote! { writer.write_all(&variant_idx.to_le_bytes()) }
    };

    Ok(quote! {
        impl #impl_generics #cratename::ser::#serialize_trait for #enum_ident #ty_generics #where_clause {
            #r#async fn serialize<#lifetime #lt_comma __W: #cratename::#writer_trait>(
                &#lifetime self,
                writer: &#lifetime mut __W,
            ) -> ::core::result::Result<(), #cratename::io::Error> {
                let variant_idx: u8 = match self {
                    #all_variants_idx_body
                };
                #write_variant_idx?;

                #fields_body
                ::core::result::Result::Ok(())
            }
        }
    })
}

fn optimize_fields_body(fields_body: TokenStream2, has_unit_variant: bool) -> TokenStream2 {
    if fields_body.is_empty() {
        // If we no variants with fields, there's nothing to match against. Just
        // re-use the empty token stream.
        fields_body
    } else {
        let unit_fields_catchall = if has_unit_variant {
            // We had some variants with unit fields, create a catch-all for
            // these to be used at the bottom.
            quote!(
                _ => {}
            )
        } else {
            TokenStream2::new()
        };
        // Create a match that serialises all the fields for each non-unit
        // variant and add a catch-all at the bottom if we do have unit
        // variants.
        quote!(
            match self {
                #fields_body
                #unit_fields_catchall
            }
        )
    }
}

#[derive(Default)]
struct VariantFields {
    header: TokenStream2,
    body: TokenStream2,
}

impl VariantFields {
    fn named_header(self) -> Self {
        let header = self.header;

        VariantFields {
            // `..` pattern matching works even if all fields were specified
            header: quote! { { #header.. }},
            body: self.body,
        }
    }
    fn unnamed_header(self) -> Self {
        let header = self.header;

        VariantFields {
            header: quote! { ( #header )},
            body: self.body,
        }
    }
}

enum VariantBody {
    // No body variant, unit enum variant.
    Unit,
    // Variant with body (fields)
    Fields(VariantFields),
}

struct VariantOutput {
    body: VariantBody,
    variant_idx_body: TokenStream2,
}

fn process_variant<const IS_ASYNC: bool>(
    variant: &Variant,
    enum_ident: &Ident,
    discriminant_value: &TokenStream2,
    cratename: &Path,
    generics: &mut serialize::GenericsOutput,
) -> syn::Result<VariantOutput> {
    let variant_ident = &variant.ident;
    let variant_output = match &variant.fields {
        Fields::Named(fields) => {
            let mut variant_fields = VariantFields::default();
            for field in &fields.named {
                let field_id = serialize::FieldId::Enum(field.ident.clone().unwrap());
                process_field::<IS_ASYNC>(
                    field,
                    field_id,
                    cratename,
                    generics,
                    &mut variant_fields,
                )?;
            }
            VariantOutput {
                body: VariantBody::Fields(variant_fields.named_header()),
                variant_idx_body: quote!(
                    #enum_ident::#variant_ident {..} => #discriminant_value,
                ),
            }
        }
        Fields::Unnamed(fields) => {
            let mut variant_fields = VariantFields::default();
            for (field_idx, field) in fields.unnamed.iter().enumerate() {
                let field_id = serialize::FieldId::new_enum_unnamed(field_idx)?;
                process_field::<IS_ASYNC>(
                    field,
                    field_id,
                    cratename,
                    generics,
                    &mut variant_fields,
                )?;
            }
            VariantOutput {
                body: VariantBody::Fields(variant_fields.unnamed_header()),
                variant_idx_body: quote!(
                    #enum_ident::#variant_ident(..) => #discriminant_value,
                ),
            }
        }
        Fields::Unit => VariantOutput {
            body: VariantBody::Unit,
            variant_idx_body: quote!(
                #enum_ident::#variant_ident => #discriminant_value,
            ),
        },
    };
    Ok(variant_output)
}

fn process_field<const IS_ASYNC: bool>(
    field: &syn::Field,
    field_id: serialize::FieldId,
    cratename: &Path,
    generics: &mut serialize::GenericsOutput,
    output: &mut VariantFields,
) -> syn::Result<()> {
    let parsed = field::Attributes::parse(&field.attrs)?;

    let needs_bounds_derive = parsed.needs_bounds_derive::<IS_ASYNC>(BoundType::Serialize);
    generics
        .overrides
        .extend(parsed.collect_bounds::<IS_ASYNC>(BoundType::Serialize));

    let field_variant_header = field_id.enum_variant_header(parsed.skip);
    if let Some(field_variant_header) = field_variant_header {
        output.header.extend(field_variant_header);
    }

    if !parsed.skip {
        let delta = field_id.serialize_output::<IS_ASYNC>(
            &field.ty,
            cratename,
            if IS_ASYNC {
                cfg_if! {
                    if #[cfg(feature = "async")] {
                        parsed.serialize_with_async
                    } else {
                        None
                    }
                }
            } else {
                parsed.serialize_with
            },
        );
        output.body.extend(delta);
        if needs_bounds_derive {
            generics.serialize_visitor.visit_field(field);
        }
    }
    Ok(())
}

#[cfg(test)]
mod tests {
    use syn::parse_quote;

    use super::*;
    use crate::internals::test_helpers::{
        default_cratename, local_insta_assert_snapshot, pretty_print_syn_str,
    };
    #[test]
    fn borsh_skip_tuple_variant_field() {
        let item_enum: ItemEnum = parse_quote! {
            enum AATTB {
                B(#[borsh(skip)] i32, #[borsh(skip)] u32),

                NegatedVariant {
                    beta: u8,
                }
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn struct_variant_field() {
        let item_enum: ItemEnum = parse_quote! {
            enum AB {
                B {
                    c: i32,
                    d: u32,
                },

                NegatedVariant {
                    beta: String,
                }
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_enum_with_custom_crate() {
        let item_enum: ItemEnum = parse_quote! {
            enum AB {
                B {
                    c: i32,
                    d: u32,
                },

                NegatedVariant {
                    beta: String,
                }
            }
        };

        let crate_: Path = parse_quote! { reexporter::borsh };

        let actual = process::<false>(item_enum.clone(), crate_.clone()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, crate_).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn borsh_skip_struct_variant_field() {
        let item_enum: ItemEnum = parse_quote! {

            enum AB {
                B {
                    #[borsh(skip)]
                    c: i32,

                    d: u32,
                },

                NegatedVariant {
                    beta: String,
                }
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn borsh_skip_struct_variant_all_fields() {
        let item_enum: ItemEnum = parse_quote! {

            enum AAB {
                B {
                    #[borsh(skip)]
                    c: i32,

                    #[borsh(skip)]
                    d: u32,
                },

                NegatedVariant {
                    beta: String,
                }
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn simple_generics() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<K, V, U> {
                B {
                    x: HashMap<K, V>,
                    y: String,
                },
                C(K, Vec<U>),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn bound_generics() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<K: Key, V, U> where V: Value {
                B {
                    x: HashMap<K, V>,
                    y: String,
                },
                C(K, Vec<U>),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn recursive_enum() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<K: Key, V> where V: Value {
                B {
                    x: HashMap<K, V>,
                    y: String,
                },
                C(K, Vec<A>),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_borsh_skip_struct_field() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<K: Key, V, U> where V: Value {
                B {
                    #[borsh(skip)]
                    x: HashMap<K, V>,
                    y: String,
                },
                C(K, Vec<U>),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_borsh_skip_tuple_field() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<K: Key, V, U> where V: Value {
                B {
                    x: HashMap<K, V>,
                    y: String,
                },
                C(K, #[borsh(skip)] Vec<U>),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn generic_serialize_bound() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<T: Debug, U> {
                C {
                    a: String,
                    #[borsh(bound(serialize =
                        "T: borsh::ser::BorshSerialize + PartialOrd,
                         U: borsh::ser::BorshSerialize"
                    ))]
                    b: HashMap<T, U>,
                },
                D(u32, u32),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn generic_serialize_async_bound() {
        let item_enum: ItemEnum = parse_quote! {
            enum A<T: Debug, U> {
                C {
                    a: String,
                    #[borsh(async_bound(serialize =
                        "T: borsh::ser::BorshSerializeAsync + PartialOrd,
                         U: borsh::ser::BorshSerializeAsync"
                    ))]
                    b: HashMap<T, U>,
                },
                D(u32, u32),
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_enum, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn check_serialize_with_attr() {
        let item_enum: ItemEnum = parse_quote! {
            enum C<K: Ord, V> {
                C3(u64, u64),
                C4 {
                    x: u64,
                    #[borsh(serialize_with = "third_party_impl::serialize_third_party")]
                    y: ThirdParty<K, V>
                },
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    #[cfg(feature = "async")]
    fn check_serialize_with_async_attr() {
        let item_enum: ItemEnum = parse_quote! {
            enum C<K: Ord, V> {
                C3(u64, u64),
                C4 {
                    x: u64,
                    #[borsh(serialize_with_async = "third_party_impl::serialize_third_party")]
                    y: ThirdParty<K, V>
                },
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        let actual = process::<true>(item_enum, default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
    }

    #[test]
    fn borsh_discriminant_false() {
        let item_enum: ItemEnum = parse_quote! {
           #[borsh(use_discriminant = false)]
            enum X {
                A,
                B = 20,
                C,
                D,
                E = 10,
                F,
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }
    #[test]
    fn borsh_discriminant_true() {
        let item_enum: ItemEnum = parse_quote! {
            #[borsh(use_discriminant = true)]
            enum X {
                A,
                B = 20,
                C,
                D,
                E = 10,
                F,
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }

    #[test]
    fn mixed_with_unit_variants() {
        let item_enum: ItemEnum = parse_quote! {
            enum X {
                A(u16),
                B,
                C {x: i32, y: i32},
                D,
            }
        };

        let actual = process::<false>(item_enum.clone(), default_cratename()).unwrap();
        local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());

        #[cfg(feature = "async")]
        {
            let actual = process::<true>(item_enum, default_cratename()).unwrap();
            local_insta_assert_snapshot!(pretty_print_syn_str(actual).unwrap());
        }
    }
}
