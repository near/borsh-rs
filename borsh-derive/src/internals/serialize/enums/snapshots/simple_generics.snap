---
source: borsh-derive/src/internals/serialize/enums/mod.rs
expression: pretty_print_syn_str(actual).unwrap()
---
impl<K, V, U> borsh::ser::BorshSerialize for A<K, V, U>
where
    K: borsh::ser::BorshSerialize,
    V: borsh::ser::BorshSerialize,
    U: borsh::ser::BorshSerialize,
{
    fn serialize<__W: borsh::io::Write>(
        &self,
        writer: &mut __W,
    ) -> ::core::result::Result<(), borsh::io::Error> {
        let variant_idx: u8 = match self {
            A::B { .. } => 0u8,
            A::C(..) => 1u8,
        };
        writer.write_all(&variant_idx.to_le_bytes())?;
        match self {
            A::B { x, y, .. } => {
                <HashMap<K, V> as borsh::BorshSerialize>::serialize(x, writer)?;
                <String as borsh::BorshSerialize>::serialize(y, writer)?;
            }
            A::C(id0, id1) => {
                <K as borsh::BorshSerialize>::serialize(id0, writer)?;
                <Vec<U> as borsh::BorshSerialize>::serialize(id1, writer)?;
            }
        }
        ::core::result::Result::Ok(())
    }
}
